### PR TITLE
JENKINS-58313 Expose source commit

### PR DESCRIPTION
**Pull request by henryju**
Relates to https://groups.google.com/d/topic/jenkinsci-dev/-ojqklsaKbw/discussion

Depends on https://github.com/jenkinsci/branch-api-plugin/pull/163

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketCommit.java [hunk 1]
```diff
@@ -48,7 +48,7 @@
     String getDate();
 
     /**
-     * @return the commit hash (complete, not reduced)
+     * @return the commit hash (short one for Bitbucket Cloud)
```
Comment (henryju) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketCommit.java]: During my tests, I noticed that Bitbucket Cloud REST API is returning a short commit sha1, not the full one. But this doesn't prevent to call other APIs using this short commit ref (for example I tested creating a build status).

File: pom.xml [hunk 2]
```diff
@@ -53,9 +53,11 @@
     <changelist>-SNAPSHOT</changelist>
     <jenkins.version>2.60.3</jenkins.version>
     <java.level>8</java.level>
-    <scm-api.version>2.2.7</scm-api.version>
+    <scm-api.version>2.4.2-SNAPSHOT</scm-api.version>
```
Comment (jglick) [inline pom.xml]: You can fix the CI errors. [Guide](https://github.com/jenkinsci/incrementals-tools/blob/master/README.md)

(Note that you will need to first bring https://github.com/jenkinsci/scm-api-plugin/pull/71 up to date with the `master` branch.)

File: pom.xml [hunk 3]
```diff
@@ -53,9 +53,11 @@
...
     <git.version>3.6.0</git.version>
     <hpi.compatibleSinceVersion>2.0</hpi.compatibleSinceVersion>
+    <!-- To fix https://github.com/mojohaus/animal-sniffer/issues/53 -->
+    <animal.sniffer.version>1.18</animal.sniffer.version>
```
Comment (jglick) [inline pom.xml]: Try updating the parent POM first.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 4]
```diff
@@ -215,8 +215,8 @@ public String getRepoOwner() {
     }
 
     @NonNull
-    public List<SCMTrait<?>> getTraits() {
-        return Collections.<SCMTrait<?>>unmodifiableList(traits);
+    public List<SCMTrait<? extends SCMTrait<?>>> getTraits() {
+        return Collections.<SCMTrait<? extends SCMTrait<?>>>unmodifiableList(traits);
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: Are these changes simply mandated by an `scm-api` update to a released version, before proceeding with your changes?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketCommit.java (hunk 1 above)
Comment (amuniz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketCommit.java]: Yes, the REST API returns the short format, but then the [API client resolves the long one](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/7c761939e6b950adbadb08ede737ac3b51470002/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java#L471) so `BitbucketCommit` instances always hold the long format.

**Comment by henryju** [comment]
See discussion on https://github.com/jenkinsci/scm-api-plugin/pull/71

**Pull request closed without merging**